File: leetcode/77.combinations.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=77 lang=java
 *
 * [77] Combinations
 */

// @lc code=start
class Solution {
    List<List<Integer>> res = new ArrayList<>();

    ArrayList<Integer> temp = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {

        dfs(1, n, k);
        return res;

    }

    public void dfs(int start, int n, int k) {
        if (temp.size() == k) {
            res.add(new ArrayList<>(temp));
        }

        for (int i = start; i < n + 1; i++) {
            temp.add(i);
            dfs(i + 1, n, k);
            temp.remove(temp.size() - 1);
        }

    }
}
// @lc code=end

File: leetcode/77.combinations.py
#
# @lc app=leetcode id=77 lang=python3
#
# [77] Combinations
#

# @lc code=start
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:

        s = []
        result = []

        def dfs(start):
            if len(s) == k:
                result.append(s[:])

            for i in range(start, n + 1):
                # if i not in s:
                #     if len(s) >= 1:
                #         if i > s[-1]:
                #             s.append(i)
                #             dfs(i + 1)
                #             s.pop()
                #     else:
                s.append(i)
                dfs(i + 1)
                s.pop()

        dfs(1)
        return result


# @lc code=end
